api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -175,6 +175,9 @@ message Cluster {
     // See the supported cluster for further documentation.
     // [#extension-category: envoy.clusters]
     google.protobuf.Any typed_config = 2;
+
+    // If true, this cluster type supports EDS updates and is configured using eds_cluster_config.
+    bool uses_eds_config = 3;
```

Comment:
I don't think this should be exposed as part of API, instead the cluster implementation should indicate that to cluster manager at C++ level.

Reply:
Agreed, this should be part of the definition of the custom cluster type.  There's no reason the control plane should need to populate a field to tell the client to do this.

Reply:
As I see it, we run into a bit of a bootstrapping issue if we do that. At the time when we decide whether a cluster should be initialized during the primary or secondary phase, we haven't yet initialized the cluster such that the cluster implementation can indicate its preference. Indeed, we don't even know if the custom cluster type even exists yet; `loadCluster()` won't reject for an unknown type until we try to initialize the cluster.

We could expose the primary/secondary status of the cluster implementation in the factory implementation and pipe that through the ClusterManagerFactory (taking care to reject an implementation that doesn't exist), calling into it before the primary phase. I'm not sure I like the idea - it's added complexity to initialization (a custom cluster now gets "pre-initialized", just enough to tell whether to initialize during primary or secondary initialization), and also means that custom clusters get initialized differently than native clusters.

An alternative I'd probably favour is simply to allow custom clusters to set `eds_cluster_config` in addition to EDS clusters, and then use `has_eds_cluster_config()` to decide whether to initialize during primary/secondary phase. I think this is still technically config-driven, though it doesn't require any changes to the config API.

Thoughts appreciated.